File: Algorithm/Python/search_in_rotated_sorted_array.py
class Solution:
    def search(self,A,target):
        if not A:
            return -1
        
        length=len(A)    
    
        l=0
        r=length-1
        while l<=r:
            mid=(l+r)/2
            print l,r,mid
            if A[mid]==target:
                return mid
            if A[mid]<A[r]: # the right is in order
                if A[mid]<target<=A[r]:
                    l=mid+1
                else:
                    r=mid-1         
            else:  # the left is in order
                if A[l]<=target<A[mid]:
                    r=mid-1
                else:
                    l=mid+1
        return -1
if __name__=='__main__':
    sol=Solution()
    A=[x for x in range(45)]
    from collections import deque
    d=deque(A)
    d.rotate(44)
    A=list(d)
    print A
    

    
    print sol.search(A,3)

File: Algorithm/CPP/search_in_rotated_sorted_array.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;
class Solution {
public:
	int search(vector<int>& nums, int target) {
		int left = 0, right = nums.size()-1;
		while (left <= right)
		{
			int mid = (right - left) / 2 + left;
			//cout << "left=" << left << " right=" << right <<" mid = "<<mid<< endl;
			if (nums[mid] == target) return mid;
			if (nums[mid]<nums[right] )  // compare the middle element with the right or the left not with the target
			{
				if (nums[mid] < target && target <= nums[right]) left = mid + 1;
				else right = mid - 1;
			}
			else
			{
				if (nums[left] <= target && target < nums[mid] ) right = mid - 1;
				else left = mid + 1;
			}
		}
		return -1;

	}
};
int  main()
{
	Solution sol;
	vector<int> nums = { 4, 5, 6, 7, 8, 1, 2, 3 };
	int target =  8;
	//nums = { 5, 1, 3 };
	//target = 3;
	cout << sol.search(nums,target) << endl;
	cin.get();
	return 0;
}
